--- src/base/mutex.cc
+++ src/base/mutex.cc
@@ -153,13 +153,7 @@ Mutex::Mutex() {
   // PTHREAD_MUTEX_RECURSIVE_NP but Mac OS X 10.5 does not
   pthread_mutexattr_t attr;
   pthread_mutexattr_init(&attr);
-#if defined(OS_MACOSX)
   pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
-#elif defined(OS_LINUX) || defined(OS_ANDROID) || defined(OS_NACL)
-  pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE_NP);
-#else
-#error "This platform is not supported."
-#endif
   pthread_mutex_init(AsPthreadMutexT(&opaque_buffer_), &attr);
 }
 
