From 22abac10b05beba3ec8126bc4765f18747cd63c7 Mon Sep 17 00:00:00 2001
From: Randy Palamar <palamar@ualberta.ca>
Date: Tue, 24 Oct 2023 06:06:03 -0600
Subject: [PATCH] support second theme and runtime swapping

---
 x.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/x.c b/x.c
index bd23686..9baca54 100644
--- a/x.c
+++ b/x.c
@@ -59,6 +59,7 @@ static void zoom(const Arg *);
 static void zoomabs(const Arg *);
 static void zoomreset(const Arg *);
 static void ttysend(const Arg *);
+static void recolor(const Arg *);
 
 /* config.h for applying patches and the configuration. */
 #include "config.h"
@@ -1917,6 +1918,17 @@ resize(XEvent *e)
 	cresize(e->xconfigure.width, e->xconfigure.height);
 }
 
+void
+recolor(const Arg *A)
+{
+	if (colorname == colordark)
+		colorname = colorlight;
+	else
+		colorname = colordark;
+	xloadcols();
+	redraw();
+}
+
 void
 run(void)
 {
-- 
2.44.0

