From ca8264b42f5d523f1e9bc9fec2f6060096a6c430 Mon Sep 17 00:00:00 2001
From: Randy Palamar <palamar@ualberta.ca>
Date: Tue, 24 Oct 2023 06:02:54 -0600
Subject: [PATCH] open in cwd

---
 config.def.h |  1 +
 st.c         | 52 +++++++++++++++++++++++++++++++++++++++++++++++++++-
 st.h         |  1 +
 x.c          |  3 +++
 4 files changed, 56 insertions(+), 1 deletion(-)

diff --git a/config.def.h b/config.def.h
index 2cd740a..897e12a 100644
--- a/config.def.h
+++ b/config.def.h
@@ -201,6 +201,7 @@ static Shortcut shortcuts[] = {
 	{ TERMMOD,              XK_Y,           selpaste,       {.i =  0} },
 	{ ShiftMask,            XK_Insert,      selpaste,       {.i =  0} },
 	{ TERMMOD,              XK_Num_Lock,    numlock,        {.i =  0} },
+	{ TERMMOD,              XK_Return,      newterm,        {.i =  0} },
 };
 
 /*
diff --git a/st.c b/st.c
index 57c6e96..bf4edfd 100644
--- a/st.c
+++ b/st.c
@@ -20,6 +20,9 @@
 #include "st.h"
 #include "win.h"
 
+extern char **g_argv;
+extern char *argv0;
+
 #if   defined(__linux)
  #include <pty.h>
 #elif defined(__OpenBSD__) || defined(__NetBSD__) || defined(__APPLE__)
@@ -28,6 +31,10 @@
  #include <libutil.h>
 #endif
 
+#if defined(__OpenBSD__)
+ #include <sys/sysctl.h>
+#endif
+
 /* Arbitrary sizes */
 #define UTF_INVALID   0xFFFD
 #define UTF_SIZ       4
@@ -796,7 +803,7 @@ ttynew(const char *line, char *cmd, const char *out, char **args)
 		if (s > 2)
 			close(s);
 #ifdef __OpenBSD__
-		if (pledge("stdio getpw proc exec", NULL) == -1)
+		if (pledge("stdio getpw proc ps exec", NULL) == -1)
 			die("pledge\n");
 #endif
 		execsh(cmd, args);
@@ -806,6 +813,7 @@ ttynew(const char *line, char *cmd, const char *out, char **args)
 		if (pledge("stdio rpath tty proc", NULL) == -1)
 			die("pledge\n");
 #endif
+		fcntl(m, F_SETFD, FD_CLOEXEC);
 		close(s);
 		cmdfd = m;
 		signal(SIGCHLD, sigchld);
@@ -1054,6 +1062,48 @@ tswapscreen(void)
 	tfulldirt();
 }
 
+static int
+chdir_by_pid(pid_t pid)
+{
+	char buf[32];
+
+#if   defined(__linux)
+	snprintf(buf, sizeof(buf), "/proc/%ld/cwd", (long)pid);
+#elif defined(__OpenBSD__)
+	size_t sz = 32;
+	int name[3] = {CTL_KERN, KERN_PROC_CWD, pid};
+	if (sysctl(name, 3, buf, &sz, 0, 0) == -1)
+		return -1;
+#endif
+	return chdir(buf);
+}
+
+void
+newterm(const Arg* a)
+{
+	int res;
+	switch (fork()) {
+	case -1:
+		die("fork failed: %s\n", strerror(errno));
+		break;
+	case 0:
+		switch (fork()) {
+		case -1:
+			die("fork failed: %s\n", strerror(errno));
+			break;
+		case 0:
+			chdir_by_pid(pid);
+			execvp("/proc/self/exe", g_argv);
+			exit(1);
+			break;
+		default:
+			exit(0);
+		}
+	default:
+		wait(NULL);
+	}
+}
+
 void
 tscrolldown(int orig, int n)
 {
diff --git a/st.h b/st.h
index fd3b0d8..f2b03b0 100644
--- a/st.h
+++ b/st.h
@@ -81,6 +81,7 @@ void die(const char *, ...);
 void redraw(void);
 void draw(void);
 
+void newterm(const Arg *);
 void printscreen(const Arg *);
 void printsel(const Arg *);
 void sendbreak(const Arg *);
diff --git a/x.c b/x.c
index 2407c7e..0f0d28c 100644
--- a/x.c
+++ b/x.c
@@ -16,6 +16,7 @@
 #include <X11/Xft/Xft.h>
 #include <X11/XKBlib.h>
 
+char **g_argv;
 char *argv0;
 #include "arg.h"
 #include "st.h"
@@ -2053,6 +2054,8 @@ main(int argc, char *argv[])
 	xw.isfixed = False;
 	xsetcursor(cursorshape);
 
+	g_argv = argv;
+
 	ARGBEGIN {
 	case 'a':
 		allowaltscreen = 0;
-- 
2.44.0

